### PR TITLE
AntelopeIO-Firewall

### PR DESCRIPTION
**Pull request by mchosc**
### Project Abstract
This is a proposal to build a firewall/proxy that would be used in an Antelope node operator's infrastructure. The end goal is to lower the barrier to entry for infrastructure providers to offer highly reliable public endpoints. We are especially excited to make it easier for anyone to operate a reliable public Antelope node without a complex understanding of DevOps.

### For which grant [level](https://github.com/eosnetworkfoundation/grant-framework#grant-levels) are you applying?

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 4 approvals

### Application Checklist

- [x] The [application template]([xxxx](https://github.com/eosnetworkfoundation/grant-framework/blob/main/applications/application-template.md)) has been copied, renamed (`my_project_name.md`) and updated.
- [x] The application file (`my_project_name.md`) has no spaces in its name, ends with the `.md` extension has been placed in the `applications` directory.
- [x] I have read and understood the:
  - [x] [FAQs](https://github.com/eosnetworkfoundation/grant-framework/blob/main/docs/faq.md), 
  - [x] [application guidelines](https://github.com/eosnetworkfoundation/grant-framework/blob/main/docs/grant_guidelines_per_category.md) and 
  - [x] [announcement guidelines](https://github.com/eosnetworkfoundation/grant-framework/blob/main/docs/announcement-guidelines.md).
- [x] An EOS address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledge the [terms and conditions](https://github.com/eosnetworkfoundation/grant-framework/blob/main/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] This project has NOT been part of a token sale or NFT sale.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/eosnetworkfoundation/grant-milestones#process).

### How Did You Hear About our grants program?

- [x] Social Media
- [ ] EOSNetwork.com
- [ ] Personal Recommendation
- [x] ENF Fireside Chat
- [ ] Investor/VC
- [ ] Online Search
- [ ] Hackathon
- [ ] Other: _______

### CODE REVIEW

**Comment by stephenpdeos** [comment]
Technical Evaluators assigned. Please note that due to the upcoming release process for Leap v4.0.0 there may be some delays. Thanks for your patience!

**Comment by tedcahalleos** [comment]
Apologies.  This grant has not been approved by the Grant Committee yet.  I jumped the gun and misinterpreted the decision data.  Please stay tuned for further feedback.

**Comment by tedcahalleos** [comment]
Hello.  The Grant Committee is currently unable to reach consensus on your application.  If it remains in this state it will be, by default, rejected.  There is one item you can choose to change that would potentially move the Grant Committee to weigh in your favor and approve the grant.  That would be agreeing to move the code to a more widely recognized systems programming language such as C, C++ or Rust.  Please let us know if you would consider moving the grant to one of those languages.  If so, please indicate that here, explain which language you will choose, update the grant, and we will have the Grant Committee review your new application at the front of their queue.  If you do not feel you can choose one of those three languages, you can indicate that here as well, and we will close the grant and mark it "Rejected".  Thank you.

**Comment by mchosc** [comment]
changed TypeScript to Rust.

**Comment by nsjames** [comment]
Sounds like a wonderful piece of software to have in the antelope arsenal. No doubt about the efficacy of the team based on their previous contributions to the ecosystem.

**Review by nsjames — APPROVED** [review]

File: applications/antelopeio_firewall.md
Comment (jdheeter) [reply applications/antelopeio_firewall.md:202]: Milestones 2 and 3 will include p2p functionality. We will be sure and clarify this.

**Comment by jdheeter** [comment]
If there is no more feedback we will start development in April. We appreciate the support and are looking forward to building this tool.

File: applications/antelopeio_firewall.md
Comment (linh2931) [inline applications/antelopeio_firewall.md:44]: Milestone 1 deliverable should contain detailed description of the static blacklist configuration formats and dynamic blacklisting algorithms.

File: applications/antelopeio_firewall.md
Comment (linh2931) [inline applications/antelopeio_firewall.md:45]: Will Milestone 1 identify exact metrics?

File: applications/antelopeio_firewall.md
Comment (linh2931) [inline applications/antelopeio_firewall.md:49]: Will a Prometheus port exposed?

**Review by linh2931 — COMMENTED** [review]
- It would be good to have a table to specify which features are implemented or initially available in which milestone.
- In Milestone 3, it is nice to have a rough measurement of delay through the firewall in a typical environment.

**Review by linh2931 — APPROVED** [review]

File: applications/antelopeio_firewall.md
Comment (jdheeter) [reply applications/antelopeio_firewall.md:44]: Based on feedback from alpha/beta users during milestone 2 development the details will become solidified. Our documentation will cover all the details.

File: applications/antelopeio_firewall.md
Comment (tedcahalleos) [reply applications/antelopeio_firewall.md]: Please clarify this for Matt now.  Not during Milestone 2 or 3.  As if Matt does not sign off, this grant will not be approved for you to proceed.  So for now, you are on hold until you give Matt the information he is seeking.

**Comment by tedcahalleos** [comment]
See my comment above.  This grant has not been approved by the Grant Evaluators.  Please address Matt's comments in detail before it is approved.

File: applications/antelopeio_firewall.md
Comment (spoonincode) [reply applications/antelopeio_firewall.md]: After some conversation with others, I'd really like to see P2P completely removed from this proposal. This is mainly for two reasons:

1. So you can focus strictly on HTTP for now, and
2. It's not clear that nodeos P2P nominally should have any sort of protection in front of it. With the upcoming P2P peer discovery feature we appear to be moving in a direction where nodes are expected to advertise their existence to each other for other nodes to connect to. This is at odds with expecting all node operators to run a P2P proxy in front of their node for safe operation.

I am ready to approve once P2P is removed.

Comment (jdheeter) [reply applications/antelopeio_firewall.md]: Ok understood

Comment (jdheeter) [reply applications/antelopeio_firewall.md]: In the latest commit I removed p2p and added some details about what features would be included in each release.

**Comment by jdheeter** [comment]
I think comments were satisfied with the latest update. Thanks for your time.

**Review by spoonincode — APPROVED** [review]

**Comment by stephenpdeos** [comment]
All technical evals complete cc @ABurton22

**Comment by jdheeter** [comment]
Hello I just wanted to bump this thread. Milestones 1 and 2 have been submitted for review @stephenpdeos @tedcahalleos

**Pull request merged**
